test/cluster/security_test.go

```diff
@@ -0,0 +1,315 @@
+		require.Eventually(t, func() bool {
+			dataEnv, err = c.Servers[follower1].QueryData(t, worldstate.DefaultDBName, strconv.Itoa(key), "admin")
+			return dataEnv != nil && dataEnv.GetResponse().GetValue() != nil && err == nil
+		}, 30*time.Second, 100*time.Millisecond)
+		dataResp := dataEnv.GetResponse().GetValue()
+		require.Equal(t, dataResp, data)
+	}
+
+	newLeaderIndex = -1
+	require.Eventually(t, func() bool {
+		newLeaderIndex = c.AgreedLeader(t, 0, 1, 2)
+		return newLeaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	require.NoError(t, c.Shutdown())
+
+	snapList = replication.ListSnapshots(clusterLogger, filepath.Join(c.Servers[newLeaderIndex].ConfigDir(), "etcdraft", "snap"))
+	require.NotEqual(t, 0, len(snapList))
+	clusterLogger.Info("Snap list: ", snapList)
+}
+
+// Scenario:
+// - start 3 servers in a cluster with TLS enabled.
+// - create server with TLS disabled, fail to start the new server.
+// - create server with TLS enabled but with a bad certificate, config tx fails.
+func TestTLSNegativeCases(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		ClusterTLSEnabled:   true,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	leaderIndex := -1
+	require.Eventually(t, func() bool {
+		leaderIndex = c.AgreedLeader(t, 0, 1, 2)
+		return leaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	// create server with TLS disabled
+	configEnv, err := c.Servers[leaderIndex].QueryConfig(t, "admin")
+	require.NoError(t, err)
+	require.NotNil(t, configEnv)
+	newServer := addServerTx(t, c, setupConfig, configEnv, leaderIndex, 3, false)
+	require.NotNil(t, newServer)
+	require.EqualError(t, c.StartServer(newServer), "failed to start the server: node-4")
+
+	// create server with TLS enabled but a bad certificate
+	t.Logf("create node-5")
+	server5, peer5, node5, err := createNewServer(c, setupConfig, 4)
+	require.NoError(t, err)
+	require.NotNil(t, server5)
+	require.NotNil(t, peer5)
+	require.NotNil(t, node5)
+	require.NoError(t, server5.CreateConfigFile(true))
+
+	configEnv, err = c.Servers[leaderIndex].QueryConfig(t, "admin")
+	require.NoError(t, err)
+	require.NotNil(t, configEnv)
+
+	node5.Certificate = []byte("invalidCert")
+	newConfig := configEnv.GetResponse().GetConfig()
+	newConfig.ConsensusConfig.Members = append(newConfig.ConsensusConfig.Members, peer5)
+	newConfig.Nodes = append(newConfig.Nodes, node5)
```

Comment:
In the local config, put a second CA from which you create a second certificate.
The server will be rejected by the cluster peers 1,2,3 and will try to communicate forever.